### PR TITLE
revisions

### PR DESCRIPTION
**Pull request by ml054**
sync with java: 3f1154f9a88a6b8b73141d55ca455a8712964d81

### CODE REVIEW

File: src/Documents/Commands/GetRevisionsCommand.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    public async setResponseAsync(bodyStream: stream.Stream, fromCache: boolean): Promise<string> {
+        if (!bodyStream) {
+            this.result = null;
+            return;
+        }
+
+        let body;
+        //TODO: view this (Greg)
+        this.result = await RavenCommandResponsePipeline.create<IRavenArrayResult>()
```
Comment (gregolsky) [inline src/Documents/Commands/GetRevisionsCommand.ts]: You can use `this._pipeline<IRavenArrayResult>()` and then the import statement is redundant.

File: src/Documents/Commands/GetRevisionsCommand.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            .collectBody(_ => body = _)
+            .parseJsonSync()
+            .streamKeyCaseTransform({
+                defaultTransform: "camel",
```
Comment (gregolsky) [inline src/Documents/Commands/GetRevisionsCommand.ts]: Is this documents coming here or something else? Sure you don't need `this._conventions.entityFieldNameConvention`?

File: src/Documents/Commands/GetRevisionsCommand.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                ignoreKeys: [ /^@/ ],
```
Comment (gregolsky) [inline src/Documents/Commands/GetRevisionsCommand.ts]: Again if these are docs you're gonna need that regex with a very long name from `CONSTANTS.METADATA` here.

File: src/Documents/Operations/Attachments/GetAttachmentOperation.ts [hunk 4]
```diff
@@ -67,27 +67,17 @@ export class GetAttachmentCommand extends RavenCommand<AttachmentResult> {
...
+            const body = this._serializer.serialize({ Type: this._type, ChangeVector: this._changeVector });
 
-        return result;
+            return {
+                uri,
+                method: "POST",
+                body
+            };
+        }
+        return { uri };
```
Comment (gregolsky) [inline src/Documents/Operations/Attachments/GetAttachmentOperation.ts:80]: double return?

File: src/Documents/Operations/Attachments/PutAttachmentOperation.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import {AttachmentData, AttachmentDetails} from "../../Attachments";
+import {IDocumentStore} from "../../IDocumentStore";
+import {DocumentConventions, RavenCommand, ServerNode} from "../../..";
+import {HttpCache} from "../../../Http/HttpCache";
+import {StringUtil} from "../../../Utility/StringUtil";
+import {throwError} from "../../../Exceptions";
+import {HttpRequestParameters} from "../../../Primitives/Http";
+import * as stream from "readable-stream";
+
+export class PutAttachmentOperation implements IOperation<AttachmentDetails> {
```
Comment (gregolsky) [inline src/Documents/Operations/Attachments/PutAttachmentOperation.ts:11]: Umm what happened to the old put attachment op? I thought it was there already...

File: src/Documents/Operations/Attachments/PutAttachmentOperation.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+        this._addChangeVectorIfNotNull(this._changeVector, req);
+
+        return req;
+    }
+
+    public async setResponseAsync(bodyStream: stream.Stream, fromCache: boolean): Promise<string> {
+        let body;
+        await this._defaultPipeline(_ => body = _).process(bodyStream)
+            .then(results => {
```
Comment (gregolsky) [inline src/Documents/Operations/Attachments/PutAttachmentOperation.ts:90]: you can do `const rawResult = await this._defaultPipeline...` then assign it to `this.result` later on. Same thing, but clearer IMO. Also you're not passing result type to `fromObjectLiteral()`

File: src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    public async setResponseAsync(bodyStream: stream.Stream, fromCache: boolean): Promise<string> {
+        if (!bodyStream) {
+            this._throwInvalidResponse();
+        }
+
+        let body;
+        await this._defaultPipeline(_ => body = _).process(bodyStream)
+            .then(results => {
+                //TODO: verify if this call is needed (Greg)
+                this.result = this._typedObjectMapper.fromObjectLiteral(results);
```
Comment (gregolsky) [inline src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts]: You do not pass types here, so it wont instantiate `ConfigureRevisionsOperationResult`. Also please add assertion in the tests for the result checking `instanceof ConfigureRevisionsOperationResult`.

File: src/Documents/Operations/RevisionsCollectionConfiguration.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import {SetupDocumentBase} from "../SetupDocumentBase";
+import {ObjectUtil} from "../../Utility/ObjectUtil";
+
+export class RevisionsCollectionConfiguration extends SetupDocumentBase {
```
Comment (gregolsky) [inline src/Documents/Operations/RevisionsCollectionConfiguration.ts:4]: While we're at it, do we need client configuration commands/operations to use`SetupDocumentBase`?

File: src/Documents/Queries/MoreLikeThis/MoreLikeThisStopWords.ts [hunk 9]
```diff
@@ -1,7 +1,12 @@
...
+import {ObjectUtil} from "../../../Utility/ObjectUtil";
 
 export class MoreLikeThisStopWords extends SetupDocumentBase {
     public id: string;
     public stopWords: string[];
-}
\ No newline at end of file
+
+    public toRemoteFieldNames() {
+        return ObjectUtil.transformObjectKeys(this, { defaultTransform: "pascal" });
```
Comment (gregolsky) [inline src/Documents/Queries/MoreLikeThis/MoreLikeThisStopWords.ts:10]: Can't recall if `transformObjectKeys()` method is mutatses the object or returning a new one ?

File: src/Documents/Session/IAttachmentsSessionOperations.ts [hunk 10]
```diff
@@ -62,8 +62,15 @@ export interface IAttachmentsSessionOperations {
...
-    //TBD AttachmentResult GetRevision(string documentId, string name, string changeVector);
-    
+    /**
+     * Returns the revision attachment by the document id and attachment name.
+     * @param documentId Document Id
+     * @param name Name of attachment
+     * @param changeVector Change vector
+     * @return Attachment
+     */
+    getRevision(documentId: string, name: string, changeVector: string): Promise<AttachmentResult>;
```
Comment (gregolsky) [inline src/Documents/Session/IAttachmentsSessionOperations.ts:71]: callback api plz?

File: src/Documents/Session/IRevisionsSessionOperations.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ * Revisions advanced session operations
+ */
+export interface IRevisionsSessionOperations {
+
+    /**
+     * Returns all previous document revisions for specified document
+     *  ordered by most recent revisions first.
+     * @param id Document id
+     */
+    getFor<TEntity extends object>(id: string): Promise<TEntity[]>;
```
Comment (gregolsky) [inline src/Documents/Session/IRevisionsSessionOperations.ts:16]: I believe we need callback api overloads for all of those

File: src/Documents/Session/Operations/GetRevisionOperation.ts [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    public getRevisionsMetadataFor(): MetadataAsDictionary[] {
+        const resultsCount = this._result.results.length;
+        const results = [] as MetadataAsDictionary[];
+
+        for (const document of this._result.results) {
+            const metadata = document[CONSTANTS.Documents.Metadata.KEY];
+            results.push(metadata);
+            //TODO: do we have to wrap this?   results.add(new MetadataAsDictionary(metadata)); (Greg)
```
Comment (gregolsky) [inline src/Documents/Session/Operations/GetRevisionOperation.ts]: question is do we need to track changes on it and I don't think so, right?

File: test/Ported/Attachments/AttachmentsRevisionsTest.ts [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+    beforeEach(async function () {
+        store = await testContext.getDocumentStore();
+    });
+
+    afterEach(async () =>
+        await disposeTestDocumentStore(store));
+
+    it("can put attachments", async () => {
+        await testContext.setupRevisions(store, false, 4);
```
Comment (gregolsky) [inline test/Ported/Attachments/AttachmentsRevisionsTest.ts:26]: can we `opts` it here? Can't tell what's false and 4.

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 1 above)
Comment (ml054) [reply src/Documents/Commands/GetRevisionsCommand.ts]: done

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 2 above)
Comment (ml054) [reply src/Documents/Commands/GetRevisionsCommand.ts]: documents wrapped with { Results: [ ..doc1, doc2]}. Is it correct then?

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 3 above)
Comment (ml054) [reply src/Documents/Commands/GetRevisionsCommand.ts]: ?

File: src/Documents/Operations/Attachments/GetAttachmentOperation.ts (hunk 4 above)
Comment (ml054) [reply src/Documents/Operations/Attachments/GetAttachmentOperation.ts:80]: I've put second return in else statement.

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 2 above)
Comment (gregolsky) [reply src/Documents/Commands/GetRevisionsCommand.ts]: nope, see `GetDocumentsCommand`

File: src/Documents/Operations/Attachments/PutAttachmentOperation.ts (hunk 5 above)
Comment (ml054) [reply src/Documents/Operations/Attachments/PutAttachmentOperation.ts:11]: You had BatchCommand, now we have  PutAttachmentOperation as second option to upload attachments.

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 3 above)
Comment (gregolsky) [reply src/Documents/Commands/GetRevisionsCommand.ts]: See `GetDocumentsCommand` :)

File: src/Documents/Operations/Attachments/PutAttachmentOperation.ts (hunk 6 above)
Comment (ml054) [reply src/Documents/Operations/Attachments/PutAttachmentOperation.ts:90]: ?

File: src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts (hunk 7 above)
Comment (ml054) [reply src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts]: ?

File: src/Documents/Commands/GetRevisionsCommand.ts (hunk 2 above)
Comment (gregolsky) [reply src/Documents/Commands/GetRevisionsCommand.ts]: https://github.com/ravendb/ravendb-nodejs-client/blob/v4.0/src/Documents/Commands/GetDocumentsCommand.ts#L205

File: src/Documents/Queries/MoreLikeThis/MoreLikeThisStopWords.ts (hunk 9 above)
Comment (ml054) [reply src/Documents/Queries/MoreLikeThis/MoreLikeThisStopWords.ts:10]: looks like it creates new object.

File: test/Ported/Attachments/AttachmentsRevisionsTest.ts (hunk 13 above)
Comment (ml054) [reply test/Ported/Attachments/AttachmentsRevisionsTest.ts:26]: just install good IDE: 

image


;-)

File: src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts (hunk 7 above)
Comment (ml054) [reply src/Documents/Operations/Revisions/ConfigureRevisionsOperation.ts]: done - thanks :)

File: src/Documents/Operations/RevisionsCollectionConfiguration.ts (hunk 8 above)
Comment (ml054) [reply src/Documents/Operations/RevisionsCollectionConfiguration.ts:4]: no - it is used only when we have props with some custom serialization logic. Like:

```
public class PutConnectionStringResult {

    @JsonProperty("ETag")   <-- in node.js we have to use SetupDocumentBase
    private Long etag;

    public Long getEtag() {
        return etag;
    }

    public void setEtag(Long etag) {
        this.etag = etag;
    }
}
```

File: src/Documents/Session/Operations/GetRevisionOperation.ts (hunk 12 above)
Comment (ml054) [reply src/Documents/Session/Operations/GetRevisionOperation.ts]: it returns revisions which are immutable, so it doesn't make sense to track changes.

Comment (ml054) [reply src/Documents/Session/Operations/GetRevisionOperation.ts]: I've removed my comment then.

File: src/Documents/Session/IAttachmentsSessionOperations.ts (hunk 10 above)
Comment (ml054) [reply src/Documents/Session/IAttachmentsSessionOperations.ts]: Please notice another method in this class doesn't have callback overrides.

Comment (ml054) [reply src/Documents/Session/IAttachmentsSessionOperations.ts]: done

Comment (ml054) [reply src/Documents/Session/IAttachmentsSessionOperations.ts:71]: done

File: src/Documents/Session/IRevisionsSessionOperations.ts (hunk 11 above)
Comment (ml054) [reply src/Documents/Session/IRevisionsSessionOperations.ts:16]: done

**Pull request merged**